ykcompile/src/lib.rs

```diff
@@ -250,41 +258,99 @@ impl TraceCompiler {
             ));
         };
 
-        // FIXME implement support for arguments.
-        if !args.is_empty() {
-            return Err(CompileError::Unimplemented("call with args".to_owned()));
+        if args.len() > 6 {
+            return Err(CompileError::Unimplemented(
+                "call with spilled args".to_owned(),
+            ));
         }
 
-        // Generally speaking, we don't know what the callee will clobber, so we save our registers
-        // before doing the call. For now we assume that the caller is OK with RAX being clobbered
-        // though.
-        for reg in 1..X86_86_NUM_REGS {
+        // Figure out where the return value (if there is one) is going.
+        let dest_reg: Option<u8> = if let Some(d) = dest {
+            Some(self.local_to_reg(d.local)?)
+        } else {
+            None
+        };
+
+        // Save Sys-V callee save registers to the stack, but skip the one (if there is one) that
+        // will store the return value. It's safe to assume the caller expects this to be
+        // clobbered.
+        //
+        // FIXME: Note that we don't save rax. Although this is a callee save register, the way the
+        // tests currently work is they check the last value returned at the end of the trace. This
+        // value is assumed to remain in rax. If we were to restore rax, we'd break that. Note that
+        // the register allocator never gives out rax for this precise reason.
+        //
+        // rdi, rsi, rdx, rcx, r8, r9, r10, r11.
+        let save_regs = [7, 6, 2, 1, 8, 9, 10, 11]
+            .iter()
+            .filter(|r| Some(**r) != dest_reg)
+            .map(|r| *r)
+            .collect::<Vec<u8>>();
+        for reg in &save_regs {
             dynasm!(self.asm
                 ; push Rq(reg)
             );
         }
 
+        let stack_index = |reg: u8| -> i32 {
+            i32::try_from(save_regs.iter().rev().position(|&r| r == reg).unwrap()).unwrap()
+        };
+
+        // Sys-V ABI dictates the first 6 arguments are passed in: rdi, rsi, rdx, rcx, r8, r9.
+        let mut arg_regs = vec![9, 8, 1, 2, 6, 7]; // reversed so they pop() in the right order.
+        for arg in args {
+            let arg_reg = if let Some(r) = arg_regs.pop() {
+                r
+            } else {
+                return Err(CompileError::Unimplemented(
+                    "out of argument registers".to_owned(),
```

Comment:
Did we not already make sure this can't happen in line 261?

Reply:
We did. I suggest we kill the one on 261.

Reply:
Isn't it better to bail earlier so we avoid extra work before erroring?

Reply:
We can do. We'd need a `panic()` or `unreachable()` here then.

Reply:
Works for me.